From 1cb0a7dea53eaccccc50ef92b1cc7a4c77dd1c5e Mon Sep 17 00:00:00 2001
From: Subrata Chatterjee <schatterjee@habana.ai>
Date: Wed, 14 Apr 2021 20:17:53 +0300
Subject: [PATCH] [SW-37130] security: flash: spi_prot() restructure

spi_prot() function is recently moved before pci is enabled.
But this added some delay before pci link is UP.
For device like goya, this delay would add further to the
already high time for the link to come up during cold boot
(eg. PRSTn).

So the spi_prot() is changed to be device specific. For
goya, the relevant function(spi_prot_late()) is called in
early preboot. For other devices(gaudi, greco), the relevant function
(spi_prot_early()) is kept before pcie link enablement in ppboot.

Change-Id: I1269a68b16075ff6b70cb2cb2cbddb90ee37b4d8
Signed-off-by: Subrata Chatterjee <schatterjee@habana.ai>
---

diff --git a/common/spi.c b/common/spi.c
index e18ecb1..389c451 100644
--- a/common/spi.c
+++ b/common/spi.c
@@ -296,6 +296,13 @@
 		preboot_puts("7\n");
 }
 
+__weak void spi_prot_early(void)
+{
+	spi_prot();
+}
+
+__weak void spi_prot_late(void) {}
+
 void spi_copy_otp(void *dst, unsigned int offset, unsigned int size)
 {
 	spi_init();
diff --git a/goya/common.c b/goya/common.c
index 8e06d3a..6a28545 100644
--- a/goya/common.c
+++ b/goya/common.c
@@ -258,3 +258,10 @@
 	memcpy(hl_ts_data, goya_hl_ts_data,
 		sizeof(struct hl_ts_cfg) * GOYA_ASIC_TS_MAX_COUNT);
 }
+
+void spi_prot_late(void)
+{
+	spi_prot();
+}
+
+void spi_prot_early(void) {}
diff --git a/include/spi.h b/include/spi.h
index 9d9564b..f55d118 100644
--- a/include/spi.h
+++ b/include/spi.h
@@ -32,5 +32,7 @@
 unsigned int get_cs_pad_sel(void);
 void spi_prot(void);
 int spi_erase(unsigned int offset, unsigned int size);
+void spi_prot_early(void);
+void spi_prot_late(void);
 
 #endif /* _HL_SPI_H_ */
diff --git a/ppboot/ppboot.c b/ppboot/ppboot.c
index 6b3eb27..f2e6661 100644
--- a/ppboot/ppboot.c
+++ b/ppboot/ppboot.c
@@ -250,7 +250,7 @@
 
 	preboot_sync_btl();
 	hl_uart_init();
-	spi_prot();
+	spi_prot_early();
 	ppboot_pcie_init();
 	ppboot_pcie_drain_init();
 	ppboot_version();
diff --git a/preboot/preboot.c b/preboot/preboot.c
index 2a7da27..756c742 100644
--- a/preboot/preboot.c
+++ b/preboot/preboot.c
@@ -31,6 +31,7 @@
 #include "i2cslave_indirect.h"
 #include "hl_i2c_slave_events.h"
 #include "rma_info.h"
+#include <spi.h>
 
 struct hl_fw_info_compat fw_info;
 
@@ -348,6 +349,8 @@
 
 	hl_uart_init();
 
+	spi_prot_late();
+
 	preboot_sp_sram_init();
 
 	/* Get the VendorID before we lock the access to configuration space */
